File: DSA/SakshamDwivedi_590013147/Day23/D23P3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

size = 0
head = None

def create_node(data):
    global head, size
    new_node = Node(data)
    if head is None:
        head = new_node
        size += 1
        return
    current = head
    while current.next:
        current = current.next
    current.next = new_node
    size += 1

def print_list():
    current = head
    while current:
        if current.next:
            print(current.data, " -> ", end="")
        else:
            print(current.data)
        current = current.next

def remove_mid():
    global head, size
    if head is None or head.next is None:
        return

    mid = size // 2
    count = 1
    trav = head

    while count != mid:
        trav = trav.next
        count += 1

    freetrav = trav.next
    trav.next = trav.next.next
    del freetrav
    size -= 1

create_node(1)
create_node(2)
create_node(3)
create_node(4)
create_node(5)
#create_node(6)

print("Before : ", end="")
print_list()

print("After : ", end="")
remove_mid()
print_list()

File: DSA/SakshamDwivedi_590013147/Day23/D23P3.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node* next;
};

int size = 0;
struct node* head = NULL;
void createNode(int);
void Print();
void RemoveMid();

int main(){
    createNode(1);
    createNode(2);
    createNode(3);
    createNode(4);
    createNode(5);
    //createNode(6);
    printf("Before : ");
    Print();
    printf("\n");
    printf("After : ");
    RemoveMid();
    Print();
}

void createNode(int data){
       struct node* Node = (struct node*)malloc(sizeof(struct node));
       Node -> data = data;
       Node -> next = NULL;
       if(head == NULL){
            head = Node;
            size++;
            return;
       }
     struct node* current = head;
     while(current -> next != NULL){
      current = current -> next;
     }
     current -> next = Node;
     size++;
}

void Print(){
    struct node* current = head;
    while(current != NULL){
        if(current -> next != NULL){
        printf("%d -> ", current -> data);}
        else{printf("%d", current -> data);}
        current = current -> next;
    }

}

void RemoveMid() {
    int mid = size / 2;
    int count = 1;
    struct node *trav = head;
    struct node *freetrav = NULL;
    while(count != mid){
        trav = trav -> next;
        count++;
    }

    freetrav = trav -> next;
    trav -> next = trav -> next -> next;
    free(freetrav);
    size--;
}
